metadata-ingestion/scripts/avro_codegen.py

```diff
@@ -333,6 +344,321 @@ class AspectBag(TypedDict, total=False):
     schema_class_file.write_text("\n".join(schema_classes_lines))
 
 
+def write_urn_classes(key_aspects: List[dict], urn_dir: Path) -> None:
+    urn_dir.mkdir()
+
+    (urn_dir / "__init__.py").write_text("\n# This file is intentionally left empty.")
+
+    code = """
+# This file contains classes corresponding to entity URNs.
+
+from typing import ClassVar, List, Optional, Type, TYPE_CHECKING
+
+import functools
+from deprecated.sphinx import deprecated as _sphinx_deprecated
+
+from datahub.utilities.urn_encoder import UrnEncoder
+from datahub.utilities.urns._urn_base import _SpecificUrn, Urn
+from datahub.utilities.urns.error import InvalidUrnError
+
+deprecated = functools.partial(_sphinx_deprecated, version="0.12.0.2")
+"""
+
+    for aspect in key_aspects:
+        entity_type = aspect["Aspect"]["keyForEntity"]
+        if aspect["Aspect"]["entityCategory"] == "internal":
+            continue
+
+        code += generate_urn_class(entity_type, aspect)
+
+    (urn_dir / "urn_defs.py").write_text(code)
+
+
+def capitalize_entity_name(entity_name: str) -> str:
+    if entity_name == "corpuser":
+        return "CorpUser"
+
+    return f"{entity_name[0].capitalize()}{entity_name[1:]}"
+
+
+def python_type(avro_type: str) -> str:
+    if avro_type == "string":
+        return "str"
+    elif (
+        isinstance(avro_type, dict)
+        and avro_type.get("type") == "enum"
+        and avro_type.get("name") == "FabricType"
+    ):
+        # TODO: make this stricter using an enum
+        return "str"
+    raise ValueError(f"unknown type {avro_type}")
+
+
+def field_type(field: dict) -> str:
+    return python_type(field["type"])
+
+
+def field_name(field: dict) -> str:
+    manual_mapping = {
+        "origin": "env",
+        "platformName": "platform_name",
+    }
+
+    name: str = field["name"]
+    if name in manual_mapping:
+        return manual_mapping[name]
+
+    # If the name is mixed case, convert to snake case.
+    if name.lower() != name:
+        return re.sub(r"(?<!^)(?=[A-Z])", "_", name).lower()
```

Comment:
Could also use the `inflection` library for this. The opportunity to use it has come up before... I think when I was writing the message for MCPDiff. I know we want to limit our dependencies but this seems pretty tame, a 500 line file with no dependencies of its own. Here's it's `underscore` impl for comparison:

```
    word = re.sub(r"([A-Z]+)([A-Z][a-z])", r'\1_\2', word)
    word = re.sub(r"([a-z\d])([A-Z])", r'\1_\2', word)
    word = word.replace("-", "_")
    return word.lower()
```
I don't really understand how the one in the code works tbh

Reply:
neither do I - it was generated by copilot :)

Reply:
actually it's basically just looking for word boundaries between lower and upper case letters and injecting an underscore in